ci, Add ci using coreos ci

RamLavi opened the pull request:
Adding a new test script that will run related coreos ci tests
with the current set-ovs.sh script
Running this before each PR will help us achive better
resiliance and ci repo integrity

Signed-off-by: Ram Lavi <ralavi@redhat.com>

RamLavi commented:
/hold
after https://github.com/coreos/coreos-assembler/pull/2221 is merged then we might want to add a test and also change the name of the current one.

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+  exit 1
+else
+  echo "tests Passed"
+fi
+
```
```suggestion
fi
```

RamLavi replied:
and Done :)

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+  echo "tests failed: ${tests_failed}"
+  exit 1
+else
+  echo "tests Passed"
```
```suggestion
  echo "tests passed"
```

RamLavi replied:
Done

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+  local image_path=$1
+  local image_url="https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/latest/latest/"
+  local latest_image_gz=$(curl ${image_url} | grep -o rhcos-[0-9].[0-9].[0-9][0-9]-x86_64-qemu.x86_64.qcow2.gz | head -1)
+  if [[ -z "${latest_image_gz}" ]]
+  then
```
```suggestion
  if [[ -z "${latest_image_gz}" ]]; then
```

Let's keep the script consistent

RamLavi replied:
Done

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+  local image_name=${latest_image_gz%.gz}
+
+  local shasum=$(curl ${image_url}/sha256sum.txt | grep ${latest_image_gz} | awk '{print $1;}')
+  if [[ -z "${shasum}" ]]
+  then
```
```suggestion
  if [[ -z "${shasum}" ]]; then
```

RamLavi replied:
Done

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+    exit 1
+  fi
+
+  if [[ $(echo "${shasum} ${image_path}/${latest_image_gz}" | sha256sum --check) != "${image_path}/${latest_image_gz}: OK" ]]
+  then
```
```suggestion
  if [[ $(echo "${shasum} ${image_path}/${latest_image_gz}" | sha256sum --check) != "${image_path}/${latest_image_gz}: OK" ]]; then
```

RamLavi replied:
Done

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+  for test_name in "${TESTS_LIST[@]}"; do
+    test_output=${TMP_COREOS_ASSEMBLER_PATH}/${test_name}_output
+    ./bin/kola run -b rhcos --qemu-image ${latest_image} ${test_name} >${test_output}
+    if [[ ! $(grep "PASS: ${test_name}" ${test_output}) ]]
+    then
```
```suggestion
    if [[ ! $(grep "PASS: ${test_name}" ${test_output}) ]]; then
```

RamLavi replied:
Done

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+
+tests_failed=$(run_tests ${latest_image})
+
+if [[ ! -z ${tests_failed} ]]
+then
```
```suggestion
if [[ ! -z ${tests_failed} ]]; then
```

RamLavi replied:
Done

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,109 @@
+
+set -ex
+
+REPO=https://github.com/coreos/coreos-assembler.git
```
Please be more specific with the variable name

RamLavi replied:
Done

phoracek reviewed (changes requested):
First iteration

RamLavi commented:
@phoracek first pass fixed [here](https://github.com/RHsyseng/rhcos-slb/compare/241b2b8a2903c1d7cd2739274375dae8f182ba5c..dc7925b5af919e77d36bea8a55547ab0c91d9285)

RamLavi commented:
@phoracek is there a way to direct a project-presubmits job to a PR instead of the main branch, so that I could check the script lane before it is merged?

phoracek commented:
> @phoracek is there a way to direct a project-presubmits job to a PR instead of the main branch, so that I could check the script lane before it is merged?

Well, a PR is just a branch on your fork, so you could setup Prow to test your fork ... but that sounds odd.

We usually setup CI for the project, so it is disabled by default and you have to trigger it with `/test`. That way it does not bother anybody.

RamLavi commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,103 @@
+TESTS_LIST=(rhcos.network.multipleNics)
+TMP_COREOS_ASSEMBLER_PATH=$(mktemp -d -u -p /tmp -t coreos-assembler-XXXXXX)
+IMAGE_PATH=/tmp/rhcos-latest-image
+RHCOS_SLB_REPO_PATH=$(pwd)
```
I'm not 100% if that would be correct when running the prow job

RamLavi commented:
updated tests on test list [here](https://github.com/RHsyseng/rhcos-slb/compare/f37c8081ee37ce306e240c100487522af50e2a9b..9b7cd0c698cfb888f307f1d6df28d459eb215bb1)

phoracek commented on tests/kola/test-coreos.sh:
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env bash
```
Any reason to keep this under kola directory? Do you anticipate adding any other tests?

RamLavi replied:
I don't. I was keeping form with kola external test convention, but this is not really an external test in cosa terms, so we can totally loose this and just put it on tests folder.

phoracek reviewed (changes requested):
I tried to run this locally and it failed with following:

```
++ for test_name in "${TESTS_LIST[@]}"
++ test_output=/tmp/coreos-assembler-TvX0PF/rhcos.network.multiple-nics_output
++ ./bin/kola run -b rhcos --qemu-image /tmp/rhcos-latest-image/rhcos-4.7.13-x86_64-qemu.x86_64.qcow2 rhcos.network.multiple-nics
Error: harness: test suite failed
```

If we turn this into CI of this repo, it should provide better failure output to give us a chance to debug it.

RamLavi commented:
> I tried to run this locally and it failed with following:
> 
> ```
> ++ for test_name in "${TESTS_LIST[@]}"
> ++ test_output=/tmp/coreos-assembler-TvX0PF/rhcos.network.multiple-nics_output
> ++ ./bin/kola run -b rhcos --qemu-image /tmp/rhcos-latest-image/rhcos-4.7.13-x86_64-qemu.x86_64.qcow2 rhcos.network.multiple-nics
> Error: harness: test suite failed
> ```
> 
> If we turn this into CI of this repo, it should provide better failure output to give us a chance to debug it.

I totally agree that we should expose this somehow.
kola tests offer a test folder with things like the test's journal.txt which is very helpful
If we use project-infra then I guess moving that folder to the artifacts folder should do it.
@kobi86 suggested running this ci on github Actions would be a better fit. It makes sense in a way, but then I need to explore how to store this output folder.

kobi86 commented:
I think we should: 
1. See if will be OK to run on Ubuntu or if we need fedora I can use my server) 
2. Create a CI for this using Github actions 
3. We can export the log file and expose it here in Github has artifact, just a step in the build:)

RamLavi commented:
every test run takes ~800Kb, so currently every PR push will add 1.6Mb
although this is a rather quiet repo, I think we may reach the free limit (500Mb) pretty fast.

RamLavi commented:
@phoracek added artifacts folder [here](https://github.com/RHsyseng/rhcos-slb/compare/9b7cd0c698cfb888f307f1d6df28d459eb215bb1..4dab594293ccb70ec6f1b0ceb64ef679967c8598)
I'll try to run it on a new fedora on beaker and see if I get anything

RamLavi commented:
@phoracek fixed artifact copy command and added prerequisites so that the tests will pass on a clean fedora [here](https://github.com/RHsyseng/rhcos-slb/compare/4dab594293ccb70ec6f1b0ceb64ef679967c8598..4c315f182939a83de6302d90c3c592186d71e54a)
Can you retry to run it locally?

RamLavi commented:
> @phoracek fixed artifact copy command and added prerequisites so that the tests will pass on a clean fedora [here](https://github.com/RHsyseng/rhcos-slb/compare/4dab594293ccb70ec6f1b0ceb64ef679967c8598..4c315f182939a83de6302d90c3c592186d71e54a)
> Can you retry to run it locally?

[update] - removed the prerequisites section to avoid running the script in sudo
Added a README section to assist local testing. 
@phoracek @kobi86 can you follow the README and tell me if it's clear&working on your local?

RamLavi commented:
> > @phoracek fixed artifact copy command and added prerequisites so that the tests will pass on a clean fedora [here](https://github.com/RHsyseng/rhcos-slb/compare/4dab594293ccb70ec6f1b0ceb64ef679967c8598..4c315f182939a83de6302d90c3c592186d71e54a)
> > Can you retry to run it locally?
> 
> [update] - removed the prerequisites section to avoid running the script in sudo
> Added a README section to assist local testing [here](https://github.com/RHsyseng/rhcos-slb/compare/4c44f6a7184dcb8452937a94a623b29986b849c5..a785055ce5b9e55fdc40635f618b19824f661f11).
> @phoracek @kobi86 can you follow the README and tell me if it's clear&working on your local?

successfully tested on clean fedora
however I'm having trouble finding `qemu` and `swtpm` when trying it on a clean rhel8

phoracek reviewed (approved):
The test has passed. The README looks good. I would not bother adding RHEL 8 specific guidelines there.

RamLavi commented on README.md:
```diff
@@ -50,6 +50,15 @@ Change *“ovs-port.bond-mode”* to the desired type and make sure that all oth
+## CI and Testing
+This repo uses [coreos-assembler repo](https://github.com/coreos/coreos-assembler) to run important scenarios relevant to this use-case.
+The test downloads the latest RHCOS image and runs network related tests, using the local `setup-ovs.sh` script.
+You can run these tests manually by running the test script:
```
it actually doesn't matter where you run the script from right now

kobi86 replied:
Cool!

kobi86 commented on README.md:
```diff
@@ -50,6 +50,15 @@ Change *“ovs-port.bond-mode”* to the desired type and make sure that all oth
 
 - MachineConfig file is a “Day 2” tool that allows to configure or run scripts on a machine with an installed OS (post-installation).
 
+## CI and Testing
```
@phoracek, I don't think we should add RHEL or other distributions to the readme, but I do think that we can add  an extra line that can make people life easy and instead wasting time to understand why it's not working they will know that this script with the command is meant to work of Fedora, if they would like to do it on other distributors then enjoy:)

RamLavi replied:
ack, [Done](https://github.com/RHsyseng/rhcos-slb/compare/a785055ce5b9e55fdc40635f618b19824f661f11..789bcdd9804bac2d0bd56c55b6bac8c7dc38fd4b)

kobi86 replied:
Just to be sure, shouldn't it be Fedora?

RamLavi replied:
totally - [Done](https://github.com/RHsyseng/rhcos-slb/compare/789bcdd9804bac2d0bd56c55b6bac8c7dc38fd4b..3858086bd73be3d684f65b84486f3b62931eadb4)

kobi86 replied:
Then LGTM:)

RamLavi commented:
@kobi86 thanks for the lgtm, can you approve it in the PR, so that we could merge it?
